src/protocols/secure_channel/MessageCounterManager.cpp

```diff
@@ -0,0 +1,344 @@
+    }
+
+    return err;
+}
+
+void MessageCounterManager::HandleMsgCounterSyncReq(Messaging::ExchangeContext * exchangeContext, const PacketHeader & packetHeader,
+                                                    System::PacketBufferHandle msgBuf)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    uint8_t * req = msgBuf->Start();
+    size_t reqlen = msgBuf->DataLength();
+
+    ChipLogDetail(SecureChannel, "Received MsgCounterSyncReq request");
+
+    VerifyOrExit(packetHeader.GetSourceNodeId().HasValue(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(req != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(reqlen == kChallengeSize, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    // Respond with MsgCounterSyncResp
+    err = SendMsgCounterSyncResp(exchangeContext, FixedByteSpan<kChallengeSize>(req));
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(SecureChannel, "Failed to handle MsgCounterSyncReq message with error:%s", ErrorStr(err));
+    }
+
+    exchangeContext->Close();
+    return;
+}
+
+void MessageCounterManager::HandleMsgCounterSyncResp(Messaging::ExchangeContext * exchangeContext,
+                                                     const PacketHeader & packetHeader, System::PacketBufferHandle msgBuf)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Transport::PeerConnectionState * state = nullptr;
+    NodeId peerNodeId                      = 0;
+    uint32_t syncCounter                   = 0;
+
+    const uint8_t * resp = msgBuf->Start();
+    size_t resplen       = msgBuf->DataLength();
+
+    ChipLogDetail(SecureChannel, "Received MsgCounterSyncResp response");
+
+    // Find an active connection to the specified peer node
+    state = mExchangeMgr->GetSessionMgr()->GetPeerConnectionState(exchangeContext->GetSecureSession());
+    VerifyOrExit(state != nullptr, err = CHIP_ERROR_NOT_CONNECTED);
+
+    VerifyOrExit(msgBuf->DataLength() == kSyncRespMsgSize, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    // Store the 64-bit value sent in the Challenge filed of the MsgCounterSyncReq.
+    VerifyOrExit(resp != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(resplen == kSyncRespMsgSize, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    syncCounter = chip::Encoding::LittleEndian::Read32(resp);
+    VerifyOrExit(syncCounter != 0, err = CHIP_ERROR_READ_FAILED);
+
+    // Verify that the response field matches the expected Challenge field for the exchange.
+    err =
+        state->GetSessionMessageCounter().GetPeerMessageCounter().VerifyChallenge(syncCounter, FixedByteSpan<kChallengeSize>(resp));
+    SuccessOrExit(err);
+
+    VerifyOrExit(packetHeader.GetSourceNodeId().HasValue(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    peerNodeId = packetHeader.GetSourceNodeId().Value();
+
+    // Process all queued ougoing and incomming group messages after message counter synchronization is completed.
```

Comment:
```suggestion
    // Process all queued incoming group messages after message counter synchronization is completed.
```
There are no outgoing messages queued...